### PR TITLE
Check for aligned chunks when writing to existing variables

### LINKED ISSUE #8371: Writing to regions with unaligned chunks can lose data
### What happened?

Writing with `region` with chunks that aren't aligned can lose data.

I've recreated an example below. While it's unlikely that folks are passing different values to `.chunk` for the template vs. the regions, I had an `"auto"` chunk, which can then set different chunk values.

(FWIW, this was fairly painful, and I managed to lose a lot of time by not noticing this, and then not really considering this could happen as I was trying to debug. I think we should really strive to ensure that we don't lose data / incorrectly report that we've successfully written data...)

### What did you expect to happen?

If there's a risk of data loss, raise an error...

### Minimal Complete Verifiable Example

```Python
ds = xr.DataArray(np.arange(120).reshape(4,3,-1),dims=list("abc")).rename('var1').to_dataset().chunk(2)

ds

# <xarray.Dataset>
# Dimensions:  (a: 4, b: 3, c: 10)
# Dimensions without coordinates: a, b, c
# Data variables:
#     var1     (a, b, c) int64 dask.array<chunksize=(2, 2, 2), meta=np.ndarray>

def write(ds):
    ds.chunk(5).to_zarr('foo.zarr', compute=False, mode='w')
    for r in (range(ds.sizes['a'])):
        ds.chunk(3).isel(a=[r]).to_zarr('foo.zarr', region=dict(a=slice(r, r+1)))


def read(ds):
    result = xr.open_zarr('foo.zarr')
    assert result.compute().identical(ds)
    print(result.chunksizes, ds.chunksizes)

write(ds); read(ds)

# AssertionError

xr.open_zarr('foo.zarr').compute()['var1']

<xarray.DataArray 'var1' (a: 4, b: 3, c: 10)>
array([[[  0,   0,   0,   3,   4,   5,   0,   0,   0,   9],
        [  0,   0,   0,  13,  14,  15,   0,   0,   0,  19],
        [  0,   0,   0,  23,  24,  25,   0,   0,   0,  29]],

       [[ 30,  31,  32,   0,   0,  35,  36,  37,  38,   0],
        [ 40,  41,  42,   0,   0,  45,  46,  47,  48,   0],
        [ 50,  51,  52,   0,   0,  55,  56,  57,  58,   0]],

       [[ 60,  61,  62,   0,   0,  65,   0,   0,   0,  69],
        [ 70,  71,  72,   0,   0,  75,   0,   0,   0,  79],
        [ 80,  81,  82,   0,   0,  85,   0,   0,   0,  89]],

       [[  0,   0,   0,  93,  94,  95,  96,  97,  98,   0],
        [  0,   0,   0, 103, 104, 105, 106, 107, 108,   0],
        [  0,   0,   0, 113, 114, 115, 116, 117, 118,   0]]])
Dimensions without coordinates: a, b, c
```

### MVCE confirmation

- [X] Minimal example — the example is as focused as reasonably possible to demonstrate the underlying issue in xarray.
- [X] Complete example — the example is self-contained, including all data and the text of any traceback.
- [X] Verifiable example — the example copy & pastes into an IPython prompt or [Binder notebook](https://mybinder.org/v2/gh/pydata/xarray/main?urlpath=lab/tree/doc/examples/blank_template.ipynb), returning the result.
- [X] New issue — a search of GitHub Issues suggests this is not a duplicate.
- [X] Recent environment — the issue occurs with the latest version of xarray and its dependencies.

### Relevant log output

_No response_

### Anything else we need to know?

_No response_

### Environment

<details>

INSTALLED VERSIONS
------------------
commit: ccc8f9987b553809fb6a40c52fa1a8a8095c8c5f
python: 3.9.18 (main, Aug 24 2023, 21:19:58)
[Clang 14.0.3 (clang-1403.0.22.14.1)]
python-bits: 64
OS: Darwin
OS-release: 22.6.0
machine: arm64
processor: arm
byteorder: little
LC_ALL: en_US.UTF-8
LANG: None
LOCALE: ('en_US', 'UTF-8')
libhdf5: None
libnetcdf: None

xarray: 2023.10.2.dev10+gccc8f998
pandas: 2.1.1
numpy: 1.25.2
scipy: 1.11.1
netCDF4: None
pydap: None
h5netcdf: None
h5py: None
Nio: None
zarr: 2.16.0
cftime: None
nc_time_axis: None
PseudoNetCDF: None
iris: None
bottleneck: None
dask: 2023.4.0
distributed: 2023.7.1
matplotlib: 3.5.1
cartopy: None
seaborn: None
numbagg: 0.2.3.dev30+gd26e29e
fsspec: 2021.11.1
cupy: None
pint: None
sparse: None
flox: None
numpy_groupies: 0.9.19
setuptools: 68.1.2
pip: 23.2.1
conda: None
pytest: 7.4.0
mypy: 1.6.0
IPython: 8.15.0
sphinx: 4.3.2

</details>

### LINKED ISSUE #8876: Possible race condition when appending to an existing zarr
### What happened?

When appending to an existing zarr along a dimension (`to_zarr(..., mode='a', append_dim="x" ,..)`), if the dask chunking of the dataset to append does not align with the chunking of the existing zarr, the resulting _consolidated_ zarr store may have `NaN`s instead of the actual values it is supposed to have.

### What did you expect to happen?

We would expected that zarr append to have the same behaviour as if we concatenate dataset _in memory_ (using `concat`) and write the whole result on a new zarr store in one go

### Minimal Complete Verifiable Example

```Python
from distributed import Client, LocalCluster
import xarray as xr
import tempfile

ds1 = xr.Dataset({"a": ("x", [1., 1.])}, coords={'x': [1, 2]}).chunk({"x": 3})
ds2 = xr.Dataset({"a": ("x", [1., 1., 1., 1.])}, coords={'x': [3, 4, 5, 6]}).chunk({"x": 3})

with Client(LocalCluster(processes=False, n_workers=1, threads_per_worker=2)): # The issue happens only when: threads_per_worker > 1
    for i in range(0, 100):
        with tempfile.TemporaryDirectory() as store:
            print(store)
            ds1.to_zarr(store, mode="w") # write first dataset
            ds2.to_zarr(store, mode="a", append_dim="x") # append first dataset

            rez = xr.open_zarr(store).compute() # open consolidated dataset
            nb_values = rez.a.count().item(0) # count non NaN values
            if nb_values != 6:
                print("found NaNs:")
                print(rez.to_dataframe())
                break
```

### MVCE confirmation

- [X] Minimal example — the example is as focused as reasonably possible to demonstrate the underlying issue in xarray.
- [X] Complete example — the example is self-contained, including all data and the text of any traceback.
- [X] Verifiable example — the example copy & pastes into an IPython prompt or [Binder notebook](https://mybinder.org/v2/gh/pydata/xarray/main?urlpath=lab/tree/doc/examples/blank_template.ipynb), returning the result.
- [X] New issue — a search of GitHub Issues suggests this is not a duplicate.
- [X] Recent environment — the issue occurs with the latest version of xarray and its dependencies.

### Relevant log output

```Python
/tmp/tmptg_pe6ox
/tmp/tmpm7ncmuxd
/tmp/tmpiqcgoiw2
/tmp/tmppma1ieo7
/tmp/tmpw5vi4cf0
/tmp/tmp1rmgwju0
/tmp/tmpm6tfswzi
found NaNs:
     a
x     
1  1.0
2  1.0
3  1.0
4  1.0
5  1.0
6  NaN
```

### Anything else we need to know?

The example code snippet provided here, reproduces the issue.

Since the issue occurs randomly, we loop in the example for a few times and stop when the issue occurs.

In the example, when `ds1` is first written, since it only contains 2 values along the `x` dimension, the resulting .zarr store have the chunking: `{'x': 2}`, even though we called `.chunk({"x": 3})`.

Side note: This behaviour in itself is not problematic in this case, but the fact that the chunking is _silently_ changed made this issue harder to spot.

However, when we try to append the second dataset `ds2`, that contains 4 values, the `.chunk({"x": 3})` in the begining splits the dask array into 2 **dask chunks**, but in a way that does not align with **zarr chunks**.

Zarr chunks:
+ chunk1 : `x: [1; 2]`
+ chunk2 : `x: [3; 4]`
+ chunk3 : `x: [5; 6]`

Dask chunks for `ds2`:
+ chunk A: `x: [3; 4; 5]`
+ chunk B: `x: [6]`

Both **dask** chunks A and B, are supposed to write on **zarr** chunk3
And depending on who writes first, we can end up with NaN on `x = 5` or `x = 6` instead of actual values.

The issue obviously happens only when dask tasks are run in parallel.
Using `safe_chunks = True` when calling `to_zarr` does not seem to help.

We couldn't figure out from the documentation how to detect this kind of issues, and how to prevent them from happening (maybe using a synchronizer?)

### Environment

<details>

INSTALLED VERSIONS
------------------
commit: None
python: 3.11.0rc1 (main, Aug 12 2022, 10:02:14) [GCC 11.2.0]
python-bits: 64
OS: Linux
OS-release: 5.15.133.1-microsoft-standard-WSL2
machine: x86_64
processor: x86_64
byteorder: little
LC_ALL: None
LANG: C.UTF-8
LOCALE: ('en_US', 'UTF-8')
libhdf5: 1.12.2
libnetcdf: 4.9.3-development

xarray: 2024.2.0
pandas: 2.2.1
numpy: 1.26.4
scipy: 1.12.0
netCDF4: 1.6.5
pydap: None
h5netcdf: 1.3.0
h5py: 3.10.0
Nio: None
zarr: 2.17.1
cftime: 1.6.3
nc_time_axis: 1.4.1
iris: None
bottleneck: 1.3.8
dask: 2024.3.1
distributed: 2024.3.1
matplotlib: 3.8.3
cartopy: None
seaborn: 0.13.2
numbagg: 0.8.1
fsspec: 2024.3.1
cupy: None
pint: None
sparse: None
flox: 0.9.5
numpy_groupies: 0.10.2
setuptools: 69.2.0
pip: 24.0
conda: None
pytest: 8.1.1
mypy: None
IPython: 8.22.2
sphinx: None

</details>

### LINKED ISSUE #8882: to_zarr silently loses data when using append_dim, if chunks are different to zarr store
### What happened?

When writing a chunked DataArray to an existing zarr store, appending along an existing dimension of the store, I have found that some data are not written if there are multiple array chunks to one zarr chunk. 

I appreciate it is probably bad practice to have different chunksizes in my DataArray and zarr_store, but I think its a realistic scenario that needs to be caught.

This may be related to / the same underlying issue as #8371. Perhaps the checks mentioned in https://github.com/pydata/xarray/issues/8371#issuecomment-1814589157 are somehow getting bypassed? Using zarr's ThreadSynchronizer is the only way I have found to ensure that all the data gets written.

### What did you expect to happen?

I expected that either 

- to_zarr would recognise the different chunk sizes, and re-chunk or wait for all the chunks to be written
- or an error would be raised, given that the results result in loss of data in an unpredictable way

### Minimal Complete Verifiable Example

```Python
import xarray as xr
import numpy as np
from matplotlib import pyplot as plt

x_coords = np.arange(10)
y_coords = np.arange(10)
t_coords = np.array([np.datetime64('2020-01-01').astype('datetime64[ns]')])
data = np.ones((10,10))

for i in range(4):
    plt.subplot(1,4,i+1)
    
    da = xr.DataArray(data.reshape((-1,10,10)),
                      dims = ['time','x','y'],
                      coords = {'x':x_coords, 'y':y_coords, 'time':t_coords},
                     ).chunk({'x':5, 'y':5,'time':1}).rename('foo')
    
    da.to_zarr('foo.zarr', mode='w')
    
    new_time = np.array([np.datetime64('2021-01-01').astype('datetime64[ns]')])
    
    da2 = xr.DataArray(data.reshape((-1,10,10)),
                      dims = ['time','x','y'],
                      coords = {'x':x_coords, 'y':y_coords, 'time':new_time},
                     ).chunk({'x':1, 'y':1,'time':1}).rename('foo')
    
    da2.to_zarr('foo.zarr',append_dim='time', mode='a')
    
    plt.imshow(xr.open_zarr('foo.zarr').isel(time=-1).foo.values)
```

### MVCE confirmation

- [ ] Minimal example — the example is as focused as reasonably possible to demonstrate the underlying issue in xarray.
- [ ] Complete example — the example is self-contained, including all data and the text of any traceback.
- [ ] Verifiable example — the example copy & pastes into an IPython prompt or [Binder notebook](https://mybinder.org/v2/gh/pydata/xarray/main?urlpath=lab/tree/doc/examples/blank_template.ipynb), returning the result.
- [ ] New issue — a search of GitHub Issues suggests this is not a duplicate.
- [ ] Recent environment — the issue occurs with the latest version of xarray and its dependencies.

### Relevant log output

_No response_

### Anything else we need to know?

Output from the plots above:

image

### Environment

<details>

INSTALLED VERSIONS
------------------
commit: None
python: 3.11.4 | packaged by conda-forge | (main, Jun 10 2023, 18:08:17) [GCC 12.2.0]
python-bits: 64
OS: Linux
OS-release: 5.15.0-1041-azure
machine: x86_64
processor: x86_64
byteorder: little
LC_ALL: C.UTF-8
LANG: C.UTF-8
LOCALE: ('en_US', 'UTF-8')
libhdf5: 1.14.3
libnetcdf: 4.9.2

xarray: 2024.2.0
pandas: 2.2.1
numpy: 1.26.4
scipy: 1.12.0
netCDF4: 1.6.5
pydap: installed
h5netcdf: 1.3.0
h5py: 3.10.0
Nio: None
zarr: 2.17.1
cftime: 1.6.3
nc_time_axis: 1.4.1
iris: None
bottleneck: 1.3.8
dask: 2024.3.1
distributed: 2024.3.1
matplotlib: 3.8.3
cartopy: 0.22.0
seaborn: 0.13.2
numbagg: None
fsspec: 2024.3.1
cupy: None
pint: 0.23
sparse: 0.15.1
flox: 0.9.5
numpy_groupies: 0.10.2
setuptools: 69.2.0
pip: 24.0
conda: 24.1.2
pytest: 8.1.1
mypy: None
IPython: 8.22.2
sphinx: None

</details>

### PR DESCRIPTION
**Pull request by max-sixty**
While I don't feel super confident that this is designed to protect against any bugs, it does solve the immediate problem in #8371, by hoisting the encoding check above the code that runs for only new variables. The encoding check is somewhat implicit, so this was an easy thing to miss prior.

- [x] Closes #8371, 
- [x] Closes #8882
- [x] Closes #8876 
- [x] Tests added
- [x] User visible changes (including notable bug fixes) are documented in `whats-new.rst`

### CODE REVIEW

File: doc/whats-new.rst [hunk 1]
```diff
@@ -64,6 +64,10 @@ Deprecations
 Bug fixes
 ~~~~~~~~~
 
+- Writing to unaligned zarr chunks with ``region`` will now raise an error
```
Comment (dcherian) [inline doc/whats-new.rst:67]:
```suggestion
- Partial writes to existing chunks with ``region`` will now raise an error
```

Didn't know what "unaligned" meant here. Does this read better?

Comment (max-sixty) [reply doc/whats-new.rst]: Thanks, that is better...

**Comment by rabernat** [comment]
@max-sixty thanks so much for tackling this tricky issue. Can you help me understand whether the following edge case is covered here...

For our existing "safe chunks" check, we have a special case for the last chunk, which does not have to evenly divide the zarr chunks:

https://github.com/pydata/xarray/blob/bb8511e0894020e180d95d2edb29ed4036ac6447/xarray/backends/zarr.py#L185-L189

(i.e. we just skip the check for the last chunk; `dchunks[:-1]`) That is appropriate because, as implemented, the assumption is that we are always writing _from the start of the array_.

With `regions` (and also with `append`), that assumption is no longer true; we now have an initial offset to consider. Imagine, for example, an array with chunksize 10, where we want to write the region `slice(5, 35)` (offset 5).  An acceptable Dask chunking would be (5, 10, 10, 5). I'm fairly certain that the existing chunk alignment code does not consider this and would not allow it. On the other hand, it _would_ allow (10, 10, 10), **which would be wrong because of the offset**!

Is handling these sorts of scenarios in scope for this PR?

**Comment by max-sixty** [comment]
Excellent point @rabernat ! Let me consider that.

@dcherian could you confirm you agree with the change assuming we handle that? I wasn't sure where your were at in #8371. TY!

**Comment by dcherian** [comment]
> Would you confirm you agree with the change assuming we handle that? 

Yes! sorry for the delay, I didn't realize that we could skip this guardrail with `safe_chunks=False`

File: xarray/backends/zarr.py [hunk 2]
```diff
@@ -679,6 +679,17 @@ def set_variables(self, variables, check_encoding_set, writer, unlimited_dims=No
             if v.encoding == {"_FillValue": None} and fill_value is None:
                 v.encoding = {}
 
+            # We need to do this for both new and existing variables to ensure we're not
+            # writing to a partial chunk, even though we don't use the `encoding` value
+            # when writing to an existing variable. See
+            # https://github.com/pydata/xarray/issues/8371 for details.
+            encoding = extract_zarr_variable_encoding(
```
Comment (dcherian) [inline xarray/backends/zarr.py:686]: Shall we specialize to calling `_determine_zarr_chunks` instead?

Comment (dcherian) [reply xarray/backends/zarr.py:709]: ah missed that this was just moved up. LGTM

**Review by dcherian — APPROVED** [review]

**Comment by dcherian** [comment]
>  we now have an initial offset to consider. Imagine, for example, an array with chunksize 10, where we want to write the region slice(5, 35) (offset 5). An acceptable Dask chunking would be (5, 10, 10, 5). I'm fairly certain that the existing chunk alignment code does not consider this and would not allow it. On the other hand, it would allow (10, 10, 10), which would be wrong because of the offset!

This is a great observation but I think we can merge this  PR for two reasons:
1. this is an improvement over the status quo
2. Users making region writes should ensure that the chunk boundaries line up. I've added a warning to the docstring for this.

Of course, it'd be nice to catch this case in the future.

**Comment by rsemlal-murmuration** [comment]
I think this adresses also, the problem of this issue: https://github.com/pydata/xarray/issues/8876

**Pull request merged**